### PR TITLE
Fix: Auto-detect chrome binary path on MacOS

### PR DESCRIPTION
**Pull request by CViniciusSDias**
On MacOS systems, the chrome binary path is different. This PR aims to fix that issue.

### CODE REVIEW

File: composer.json [hunk 1]
```diff
@@ -15,7 +15,7 @@
         }
     ],
     "require": {
-        "php": "^7.1.3 || ^8.0",
+        "php": ">=7.2",
```
Comment (GrahamCampbell) [inline composer.json]: Please revert this change.

File: tests/BrowserFactoryTest.php [hunk 2]
```diff
@@ -26,7 +26,7 @@ public function testBrowserFactory()
 
         $browser = $factory->createBrowser();
 
-        $this->assertRegExp('#^ws://#', $browser->getSocketUri());
+        $this->assertMatchesRegularExpression('#^ws://#', $browser->getSocketUri());
```
Comment (GrahamCampbell) [inline tests/BrowserFactoryTest.php]: Please don;t change this.

**Review by GrahamCampbell — CHANGES_REQUESTED** [review]

File: tests/BrowserFactoryTest.php (hunk 2 above)
Comment (CViniciusSDias) [reply tests/BrowserFactoryTest.php]: `assertRegExp` is deprecated. Why not to change it?

File: composer.json (hunk 1 above)
Comment (CViniciusSDias) [reply composer.json]: `PHP_OS_FAMILY` constant is available from PHP 7.2 up. I can change the code to use `PHP_OS`. I'll work on that tonight.

Comment (GrahamCampbell) [reply composer.json]: Pull in the symfony php 7.2 polyfill, please.

File: tests/BrowserFactoryTest.php (hunk 2 above)
Comment (GrahamCampbell) [reply tests/BrowserFactoryTest.php]: Because it won;t work on php 7.1.

**Comment by enricodias** [comment]
If we're going to add a feature to auto discover the chrome binary based on the OS, wouldn't it be better to have a separate method to do the discovery instead of the current `setChromeBinary`? This method is doing 2 things.

Maybe something like this would be better:

```php
    public function __construct(string $chromeBinary = null)
    {
        $this->chromeBinary = $chromeBinary ?? $this->getDefaultChromeBinaryPath();
    }

    private function getDefaultChromeBinaryPath(): void
    {
        if (array_key_exists('CHROME_PATH', $_SERVER)) {
            return $_SERVER['CHROME_PATH'];
        }

        if (PHP_OS_FAMILY === 'Darwin') {
           return '/Applications/Google Chrome.app/Contents/MacOS/Google Chrome';
        }

        return 'chrome';
    }
```

The method could be modified to work with Windows later on, or even query the OS to find the correct path.

**Comment by CViniciusSDias** [comment]
> The method could be modified to work with Windows later on, or even query the OS to find the correct path.

AFAIK, `chrome` works fine on _Windows_

File: tests/BrowserFactoryTest.php (hunk 2 above)
Comment (CViniciusSDias) [reply tests/BrowserFactoryTest.php]: I checked PHPUnit's code and I did not find anything that would make `assertMatchesRegularExpression` fail on PHP 7.1.

I'm having some docker problems so I'm not being able to reproduce the environment. Because of that I'll revert this change, but I still don't understand what would break on PHP 7.1

File: composer.json (hunk 1 above)
Comment (CViniciusSDias) [reply composer.json]: I used `PHP_OS` since we have the same result without an extra dependency.

**Review by GrahamCampbell — CHANGES_REQUESTED** [review]
Pull in the symfony php 7.2 polyfill, please.

**Comment by CViniciusSDias** [comment]
> Pull in the symfony php 7.2 polyfill, please.

What's the reason for that? If it's just for `PHP_OS_FAMILY` I'd rather not since for MacOS the result is exactly the same.

**Comment by GrahamCampbell** [comment]
Oh yes, I didn't spot that change.

**Comment by enricodias** [comment]
> AFAIK, `chrome` works fine on _Windows_

It doesn't, at least on my machines. I have to use the full path.

**Comment by GrahamCampbell** [comment]
What's the most common path to 64-bit chrome on Windows? We might as well throw that in here too.

**Comment by CViniciusSDias** [comment]
> What's the most common path to 64-bit chrome on Windows? We might as well throw that in here too.

Apparently `C:\Program Files (x86)\Google\Chrome\Application\Chrome.exe`. Could you please confirm that, @enricodias ?

**Comment by enricodias** [comment]
@GrahamCampbell it varies a lot. On my current machine it's in `Program Files (x86)\Google\Chrome\Application\chrome.exe` but the "Program Files" folder itself gets translated depending on the installed language. The full path also varies between different Windows versions.

It's probably better to find the correct path using the Windows registry somehow.

**Comment by CViniciusSDias** [comment]
The "Program Files" folder only has alias in the Explorer, but it's always "Program Files" from the CLI. About Windows versions, I don't believe that changes...

**Comment by enricodias** [comment]
@CViniciusSDias On Windows 7 it's in `C:\Program Files (x86)\Google\Application\chrome.exe`, on Vista I believe it was in `C:\Users\UserName\AppDataLocal\Google\Chrome`. I also think that I've seen it in `LocalAppData` before aswell.

Using "Program Files (x86)" worked here in the CLI, but you'll also need to guess the drive letter used. Mine for example wasn't using the default `c:/`

**Comment by CViniciusSDias** [comment]
Ah, I didn't take into consideration the possibility of using something different than `C:`.
Do you know how to get the path from the registry, @enricodias ?
If not, this will probably need some further investigation and I believe it would be better to have it in a different PR, @GrahamCampbell

**Comment by CViniciusSDias** [comment]
Also, I believe we'd need an extension to get values from Windows registry via PHP...

**Comment by enricodias** [comment]
Using the env variable `%ProgramFiles(x86)%` also worked here in the old cmd, but it doesn't return the correct drive letter.

I'll have to take a look on how to query the registry using php without admin permissions. I'm not 100% sure if it's possible.

**Comment by enricodias** [comment]
`start chrome` also worked in the cmd, but it doesn't work as a path to start the process.

I found the [amphp/windows-registry](https://github.com/amphp/windows-registry) package that seems to be able to read keys from the registry. I'll test it later.

**Comment by GrahamCampbell** [comment]
> On Windows 7 it's in C:\Program Files (x86)\Google\Application\chrome.exe

That is not 64-bit chrome?

**Comment by enricodias** [comment]
> That is not 64-bit chrome?

Even though the chrome download page says 64bit, the automated installer always used the x86 version for me. The installer itself doesn't show any options during the setup. That's one more thing to guess; even if the user is using a 64bit system, chrome could be x86.

**Comment by GrahamCampbell** [comment]
Thanks.
